### PR TITLE
Datafusion upstream merge

### PR DESCRIPTION
**Pull request by jdye64**
Merge upstream/main into the datafusion-sql-planner branch after the recent release.

### CODE REVIEW

File: dask_sql/mappings.py [hunk 1]
```diff
@@ -283,15 +283,17 @@ def cast_column_to_type(col: dd.Series, expected_type: str):
...
-    current_float = pd.api.types.is_float_dtype(current_type)
-    expected_integer = pd.api.types.is_integer_dtype(expected_type)
-    if current_float and expected_integer:
-        logger.debug("...truncating...")
-        # Currently "trunc" can not be applied to NA (the pandas missing value type),
-        # because NA is a different type. It works with np.NaN though.
-        # For our use case, that does not matter, as the conversion to integer later
-        # will convert both NA and np.NaN to NA.
-        col = da.trunc(col.fillna(value=np.NaN))
+    if pd.api.types.is_integer_dtype(expected_type):
```
Comment (ayushdg) [inline dask_sql/mappings.py:286]: Some of these changes were added to incorporate the `timestampdiff` related operations

File: dask_sql/physical/rex/core/call.py [hunk 2]
```diff
@@ -169,11 +169,10 @@ def div(self, lhs, rhs):
         # We do not need to truncate in this case
         # So far, I did not spot any other occurrence
         # of this function.
-        if isinstance(result, datetime.timedelta):
-            return result
-        else:  # pragma: no cover
-            result = da.trunc(result)
+        if isinstance(result, (datetime.timedelta, np.timedelta64)):
```
Comment (ayushdg) [inline dask_sql/physical/rex/core/call.py:172]: Same

File: dask_sql/physical/rex/core/call.py [hunk 3]
```diff
@@ -818,6 +814,7 @@ class RexCallPlugin(BaseRexPlugin):
         "/int": IntDivisionOperator(),
         # special operations
         "cast": CastOperation(),
+        "reinterpret": CastOperation(),
```
Comment (ayushdg) [inline dask_sql/physical/rex/core/call.py:817]: Same

**Review by ayushdg — CHANGES_REQUESTED** [review]
Given that some of the changes here are ones that were added for timestampdiff. It might make sense to exclude all the changes for now that were made in #495 apart from tests or include everything and remove the parts that are not relevant later.

**Comment by jdye64** [comment]
@ayushdg Does the last commit address what you were talking about?

**Comment by ayushdg** [comment]
> @ayushdg Does the last commit address what you were talking about?

It does but I now realize it's still missing some of the old logic that was updated for timestampdiff, eg places like [this]( https://github.com/dask-contrib/dask-sql/pull/495/files#diff-d3cd58a1b69c1ecccbefe8b23cd8e3c78cfc005841dd6711a094e75f388a289eL290).


To ensure that some of these changes don't slip through, I'm actually considering just including all the changes made to #495 and then removing bit that are not needed in a followup. Apologies if that increases the amount work needed on this pr.

**Review by ayushdg — APPROVED** [review]
Some changes from #581 also made it in. But changes lgtm!

**Pull request merged**